misc/python/materialize/cli/ci_coverage_pr_report.py

```diff
@@ -95,41 +111,100 @@ def mark_covered_lines(lcov_file: str, coverage: Coverage) -> None:
             line_nr = int(line_str)
             hit = int(hit_str) if hit_str.isnumeric() else int(float(hit_str))
             if line_nr in coverage[file]:
-                coverage[file][line_nr] = (coverage[file][line_nr] or 0) + hit
+                if unittests:
+                    if not coverage[file][line_nr]:
+                        coverage[file][line_nr] = (coverage[file][line_nr] or 0) - hit
+                else:
+                    coverage[file][line_nr] = (coverage[file][line_nr] or 0) + hit
 
 
 def get_report(coverage: Coverage) -> str:
     """
     Remove uncovered lines in real files and print a git diff, then restore to
-    original state.
+    original state. This is pretty messy, we might want to do it without git if
+    it gets any more complex.
     """
     try:
+        # Remove lines which are only covered in unit tests
         for file, lines in coverage.items():
             with open(file, "r+") as f:
                 content = f.readlines()
                 f.seek(0)
                 for i, line in enumerate(content):
-                    if lines.get(i + 1) != 0:
+                    if (lines.get(i + 1) or 0) < 0:
+                        f.write(line)
+                f.truncate()
+
+        subprocess.run(
+            ["git", "config", "user.email", "coverage@materialize.com"],
+            check=True,
+        )
+        subprocess.run(
+            ["git", "config", "user.name", "Code Coverage"],
+            check=True,
+        )
+        subprocess.run(
+            ["git", "commit", "--allow-empty", "-a", "-m", "Covered in unit test only"],
+            check=True,
+        )
+        # Add the lines back so they show up with a "." marker
+        subprocess.run(["git", "revert", "--no-commit", "HEAD"], check=True)
+
+        # Remove lines which are not covered at all so they show up with "!" marker
+        for file, lines in coverage.items():
+            with open(file, "r+") as f:
+                content = f.readlines()
+                f.seek(0)
+                for i, line in enumerate(content):
+                    if (lines.get(i + 1) or 0) > 0:
                         f.write(line)
                 f.truncate()
 
         result = subprocess.run(
-            ["git", "diff", "--output-indicator-old=!"], check=True, capture_output=True
+            [
+                "git",
+                "diff",
+                "--output-indicator-old=!",
```

Comment:
I think this would be the final form of this line of development, as there are no more output indicators to be had :-) 

If this needs to get even more complex, I think you should consider refactoring the entire operation around what can be called LineSets, which is a list of files+lines from any source, e.g. the complete source, a single coverage report or a patch diff.

In this case, the final reports can be expressed as set operations, unions and intersections over such LineSets:

```
coverage_from_tests = (coverage_from_test1 UNION ... UNION coverage_from_testN)
coverage_from_all = coverage_from_tests UNION coverage_from_unit_tests

lines_not_covered_in_test = patch EXCEPT coverage_from_tests
lines_only_covered_by_unit_tests = coverage_from_all EXCEPT coverage_from_test

report_lines_not_covered_in_test = DIFF_BETWEEN(entire_codebase AND coverage_from_tests)
report_lines_covered_only_by_unit_tests = DIFF_BETWEEN(entire_codebase AND lines_only_covered_by_unit_tests)
```

or something along those lines, you get the idea.

The benefit of having such algebra over LineSets , is, (apart form keeping @aalexandrov happy that everything that is not a lattice is an algebra) that it can be used to perform more complex calculations over coverage reports. For example, in previous companies I experimented with the following two things:

- with randomly generated queries, collected a per-query coverage report. With algebra operations over coverage reports one can easily implement a script that calculates the unique contribution that each query brings to the entire coverage, order the queries by their contribution and pick the N top-most queries so that the total coverage provided by those N queries was equivalent to the coverage provided by the entire starting set of queries. Those N queries can then be made into an SLT test and pushed into the codebase.
- with test case minimization tools, such a sqlreducer, one can get a separate coverage report for all queries that showed the problem while it was being reduced and a separate report for all queries that did not show the problem ever. Subtracting one report from the other gives us a set of code lines that are common for all failing queries but never occur in successful queries. Sometimes this report narrows down the difference between success and failures to a specific if statement or a code block that is always taken for failure to happen.